File: startrek-java/StarTrek/src/main/java/chat/dim/net/BaseHub.java
/* license: https://mit-license.org
 *
 *  Star Trek: Interstellar Transport
 *
 *                                Written in 2020 by Moky <albert.moky@gmail.com>
 *
 * ==============================================================================
 * The MIT License (MIT)
 *
 * Copyright (c) 2020 Albert Moky
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * ==============================================================================
 */
package chat.dim.net;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.net.SocketAddress;
import java.nio.ByteBuffer;
import java.util.Set;

import chat.dim.type.AddressPairMap;

class ConnectionPool extends AddressPairMap<Connection> {

    @Override
    public void set(SocketAddress remote, SocketAddress local, Connection value) {
        Connection old = get(remote, local);
        if (old != null && old != value) {
            remove(remote, local, old);
        }
        super.set(remote, local, value);
    }

    @Override
    public Connection remove(SocketAddress remote, SocketAddress local, Connection value) {
        Connection cached = super.remove(remote, local, value);
        if (cached != null && cached.isOpen()) {
            cached.close();
        }
        return cached;
    }
}

public abstract class BaseHub implements Hub {

    /*  Maximum Segment Size
     *  ~~~~~~~~~~~~~~~~~~~~
     *  Buffer size for receiving package
     *
     *  MTU        : 1500 bytes (excludes 14 bytes ethernet header & 4 bytes FCS)
     *  IP header  :   20 bytes
     *  TCP header :   20 bytes
     *  UDP header :    8 bytes
     */
    public static int MSS = 1472;  // 1500 - 20 - 8

    private final ConnectionPool connectionPool = new ConnectionPool();
    private final WeakReference<Connection.Delegate> delegateRef;

    protected BaseHub(Connection.Delegate delegate) {
        super();
        delegateRef = new WeakReference<>(delegate);
    }

    public Connection.Delegate getDelegate() {
        return delegateRef.get();
    }

    /**
     *  Get all channels
     *
     * @return copy of channels
     */
    protected abstract Set<Channel> allChannels();

    /**
     *  Remove socket channel
     *
     * @param channel - socket channel
     */
    protected abstract void removeChannel(Channel channel);

    /**
     *  Create connection with sock channel & addresses
     *
     * @param sock   - socket channel
     * @param remote - remote address
     * @param local  - local address
     * @return null on channel not exists
     */
    protected abstract Connection createConnection(Channel sock, SocketAddress remote, SocketAddress local);

    protected Set<Connection> allConnections() {
        return connectionPool.allValues();
    }
    protected Connection getConnection(SocketAddress remote, SocketAddress local) {
        return connectionPool.get(remote, local);
    }
    protected void putConnection(Connection conn) {
        connectionPool.set(conn.getRemoteAddress(), conn.getLocalAddress(), conn);
    }
    protected void removeConnection(Connection conn) {
        connectionPool.remove(conn.getRemoteAddress(), conn.getLocalAddress(), conn);
    }

    @Override
    public Connection connect(SocketAddress remote, SocketAddress local) {
        Connection conn = getConnection(remote, local);
        if (conn != null) {
            // check local address
            if (local == null) {
                return conn;
            }
            SocketAddress address = conn.getLocalAddress();
            if (address == null || address.equals(local)) {
                return conn;
            }
            // local address not matched? ignore this connection
        }
        // try to open channel with direction (remote, local)
        Channel sock = open(remote, local);
        if (sock == null || !sock.isOpen()) {
            return null;
        }
        // create with channel
        conn = createConnection(sock, remote, local);
        if (conn != null) {
            // NOTICE: local address in the connection may be set to None
            putConnection(conn);
        }
        return conn;
    }

    protected boolean driveChannel(Channel sock) {
        ByteBuffer buffer = ByteBuffer.allocate(MSS);
        SocketAddress remote;
        // try to receive
        try {
            remote = sock.receive(buffer);
        } catch (IOException error) {
            //e.printStackTrace();
            removeChannel(sock);
            // callback
            Connection.Delegate delegate = getDelegate();
            if (delegate != null) {
                remote = sock.getRemoteAddress();
                SocketAddress local = sock.getLocalAddress();
                delegate.onError(error, null, remote, local, null);
            }
            return false;
        }
        if (remote == null) {
            // received nothing
            return false;
        }
        SocketAddress local = sock.getLocalAddress();
        // get connection for processing received data
        Connection conn = connect(remote, local);
        if (conn != null) {
            byte[] data = new byte[buffer.position()];
            buffer.flip();
            buffer.get(data);
            conn.received(data, remote, local);
        }
        return true;
    }
    protected int driveChannels(Set<Channel> channels) {
        int count = 0;
        for (Channel sock : channels) {
            if (sock.isAlive() && driveChannel(sock)) {
                // received data from this socket channel
                count += 1;
            }
        }
        return count;
    }
    protected void cleanupChannels(Set<Channel> channels) {
        for (Channel sock : channels) {
            if (!sock.isAlive()) {
                removeChannel(sock);
            }
        }
    }

    protected void driveConnections(Set<Connection> connections) {
        for (Connection conn : connections) {
            try {
                // drive connection to go on
                conn.tick();
            } catch (Throwable e) {
                e.printStackTrace();
            }
            // NOTICE: let the delegate to decide whether close an error connection
            //         or just remove it.
        }
    }
    protected void cleanupConnections(Set<Connection> connections) {
        for (Connection conn : connections) {
            if (!conn.isAlive()) {
                removeConnection(conn);
            }
        }
    }

    @Override
    public boolean process() {
        try {
            // 1. drive all channels to receive data
            Set<Channel> channels = allChannels();
            int count = driveChannels(channels);
            // 2. drive all connections to move on
            Set<Connection> connections = allConnections();
            driveConnections(connections);
            // 3. cleanup closed channels and connections
            cleanupChannels(channels);
            cleanupConnections(connections);
            return count > 0;
        } catch (Throwable e) {
            e.printStackTrace();
            return false;
        }
    }
}

File: startrek-py/startrek/net/base_hub.py
# -*- coding: utf-8 -*-
#
#   Star Trek: Interstellar Transport
#
#                                Written in 2021 by Moky <albert.moky@gmail.com>
#
# ==============================================================================
# MIT License
#
# Copyright (c) 2021 Albert Moky
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# ==============================================================================

import socket
import traceback
import weakref
from abc import ABC, abstractmethod
from typing import Optional, Set

from ..types import Address, AddressPairMap

from .hub import Hub
from .channel import Channel
from .connection import Connection
from .delegate import ConnectionDelegate


class ConnectionPool(AddressPairMap[Connection]):

    # noinspection PyMethodMayBeStatic
    def _close_connection(self, connection: Connection):
        if connection.opened:
            connection.close()

    # Override
    def set(self, remote: Optional[Address], local: Optional[Address], item: Optional[Connection]):
        old = self.get(remote=remote, local=local)
        if old is not None and old is not item:
            self._close_connection(connection=old)
        super().set(remote=remote, local=local, item=item)

    # Override
    def remove(self, remote: Optional[Address], local: Optional[Address],
               item: Optional[Connection]) -> Optional[Connection]:
        cached = super().remove(remote=remote, local=local, item=item)
        if cached is not None:
            self._close_connection(connection=cached)
            return cached


class BaseHub(Hub, ABC):

    """
        Maximum Segment Size
        ~~~~~~~~~~~~~~~~~~~~
        Buffer size for receiving package

        MTU        : 1500 bytes (excludes 14 bytes ethernet header & 4 bytes FCS)
        IP header  :   20 bytes
        TCP header :   20 bytes
        UDP header :    8 bytes
    """
    MSS = 1472  # 1500 - 20 - 8

    def __init__(self, delegate: ConnectionDelegate):
        super().__init__()
        self.__delegate = weakref.ref(delegate)
        self.__connection_pool = ConnectionPool()

    @property
    def delegate(self) -> ConnectionDelegate:
        return self.__delegate()

    @abstractmethod
    def _all_channels(self) -> Set[Channel]:
        """
        get all channels

        :return: copy of channels
        """
        raise NotImplemented

    @abstractmethod
    def _remove_channel(self, channel: Channel):
        """
        remove socket channel

        :param channel: socket channel
        """
        raise NotImplemented

    @abstractmethod
    def _create_connection(self, channel: Channel, remote: Address, local: Optional[Address]) -> Optional[Connection]:
        """
        create connection with channel channel & addresses

        :param channel: channel channel
        :param remote:  remote address
        :param local:   local address
        :return: None on channel not exists
        """
        raise NotImplemented

    def _all_connections(self) -> Set[Connection]:
        """ get a copy of connections """
        return self.__connection_pool.items

    def _get_connection(self, remote: Optional[Address], local: Optional[Address]) -> Optional[Connection]:
        """ get cached connection """
        return self.__connection_pool.get(remote=remote, local=local)

    def _set_connection(self, connection: Connection):
        """ cache connection """
        self.__connection_pool.set(remote=connection.remote_address, local=connection.local_address, item=connection)

    def _remove_connection(self, connection: Connection):
        """ remove cached connection """
        self.__connection_pool.remove(remote=connection.remote_address, local=connection.local_address, item=connection)

    # Override
    def connect(self, remote: Address, local: Optional[Address] = None) -> Optional[Connection]:
        conn = self._get_connection(remote=remote, local=local)
        if conn is not None:
            # check local address
            if local is None:
                return conn
            address = conn.local_address
            if address is None or address == local:
                return conn
            # local address not matched? ignore this connection
        # try to open channel with direction (remote, local)
        channel = self.open(remote=remote, local=local)
        if channel is None or not channel.opened:
            return None
        # create with channel
        conn = self._create_connection(channel=channel, remote=remote, local=local)
        if conn is not None:
            # cache connection for (remote, local)
            self._set_connection(connection=conn)
            return conn

    def _drive_channel(self, channel: Channel) -> bool:
        local = channel.local_address
        # try to receive
        try:
            data, remote = channel.receive(max_len=self.MSS)
        except socket.error as error:
            # print('[NET] failed to receive data: %s' % error)
            remote = channel.remote_address
            # socket error, close the channel
            channel.close()
            # callback
            delegate = self.delegate
            if delegate is not None:
                delegate.connection_error(error=error, data=None, source=remote, destination=local, connection=None)
            return False
        if remote is None:
            # received nothing
            return False
        # get connection for processing received data
        conn = self.connect(remote=remote, local=local)
        if conn is not None:
            conn.received(data=data, remote=remote, local=local)
        return True

    def _drive_channels(self, channels: Set[Channel]) -> int:
        count = 0
        for sock in channels:
            try:
                if sock.alive and self._drive_channel(channel=sock):
                    count += 1  # received data from this socket channel
            except Exception as error:
                print('[NET] drive channel error: %s, %s' % (error, sock))
                traceback.print_exc()
        return count

    def _drive_connections(self, connections: Set[Connection]):
        for conn in connections:
            try:
                conn.tick()  # drive connection to go on
            except Exception as error:
                print('[NET] drive connection error: %s, %s, %s' % (error, conn, self))
                traceback.print_exc()
            # NOTICE: let the delegate to decide whether close an error connection
            #         or just remove it.

    def _cleanup_channels(self, channels: Set[Channel]):
        for sock in channels:
            if not sock.alive:
                self._remove_channel(channel=sock)

    def _cleanup_connections(self, connections: Set[Connection]):
        # NOTICE: multi connections may share same channel (UDP Hub)
        for conn in connections:
            if not conn.opened:
                self._remove_connection(connection=conn)

    # Override
    def process(self) -> bool:
        try:
            # 1. drive all channels to receive data
            channels = self._all_channels()
            count = self._drive_channels(channels=channels)
            # 2. drive all connections to move on
            connections = self._all_connections()
            self._drive_connections(connections=connections)
            # 3. cleanup closed channels and connections
            self._cleanup_channels(channels=channels)
            self._cleanup_connections(connections=connections)
            return count > 0
        except Exception as error:
            print('[NET] hub process error: %s' % error)
            traceback.print_exc()
